snapcraft/internal/repo.py

```diff
@@ -198,106 +189,85 @@ def _setup_apt(self, download_dir):
-            apt_cache_dir, 'etc', 'apt', 'sources.list')
+            cache_dir, 'etc', 'apt', 'sources.list')
 
         os.makedirs(os.path.dirname(sources_list_file), exist_ok=True)
         with open(sources_list_file, 'w') as f:
-            f.write(sources_list)
+            f.write(self._collected_sources_list())
 
         # dpkg also needs to be in the rootdir in order to support multiarch
         # (apt calls dpkg --print-foreign-architectures).
         dpkg_path = shutil.which('dpkg')
         if dpkg_path:
             # Symlink it into place
-            destination = os.path.join(apt_cache_dir, dpkg_path[1:])
+            destination = os.path.join(cache_dir, dpkg_path[1:])
             if not os.path.exists(destination):
                 os.makedirs(os.path.dirname(destination), exist_ok=True)
                 os.symlink(dpkg_path, destination)
         else:
             logger.warning(
                 "Cannot find 'dpkg' command needed to support multiarch")
 
-        apt_cache = apt.Cache(rootdir=apt_cache_dir, memonly=True)
+        apt_cache = apt.Cache(rootdir=cache_dir, memonly=True)
         apt_cache.update(fetch_progress=self.progress,
                          sources_list=sources_list_file)
 
-        copy_tree(apt_cache_dir, rootdir, update=True)
-
-        return package_cache_dir
+        return apt_cache
 
-    def _restore_cached_packages(self, apt_changes,
-                                 package_cache_dir, download_dir):
-        for pkg in apt_changes:
-            src = os.path.join(package_cache_dir, pkg.name)
-            dst = os.path.join(download_dir, pkg.name)
-            if os.path.exists(src):
-                file_utils.link_or_copy(src, dst)
-
-    def _store_cached_packages(self, package_cache_dir, download_dir):
-        os.makedirs(package_cache_dir, exist_ok=True)
-        for pkg in os.listdir(download_dir):
-            if not pkg.endswith('.deb'):
-                continue
-            src = os.path.join(download_dir, pkg)
-            dst = os.path.join(package_cache_dir, pkg)
-            # The dst may be an incomplete or broken so let's update
-            # just in case.
-            if os.path.exists(dst):
-                os.unlink(dst)
-            file_utils.link_or_copy(src, dst)
-
-    @contextmanager
-    def archive(self, rootdir, download_dir):
+    @contextlib.contextmanager
+    def archive(self, cache_dir):
         try:
-            self._setup_apt(download_dir)
-            package_cache_dir = self._setup_apt_cache(rootdir)
-            apt_cache = apt.Cache(rootdir=rootdir, memonly=True)
+            apt_cache = self._setup_apt(cache_dir)
             apt_cache.open()
-            self._restore_cached_packages(apt_cache.get_changes(),
-                                          package_cache_dir, download_dir)
+
             yield apt_cache
-            self._store_cached_packages(package_cache_dir, download_dir)
+
+            apt_cache.close()
```

Comment:
what if an exception occurs during the `yield`, you should jump to `finally` and `.close` there which will happen no matter what